nbconvert/preprocessors/execute.py

```diff
@@ -473,49 +501,77 @@ def _wait_for_reply(self, msg_id, cell=None):
                 if timeout and cummulative_time > timeout:
-                    self.log.error(
-                        "Timeout waiting for execute reply (%is)." % self.timeout)
-                    if self.interrupt_on_timeout:
-                        self.log.error("Interrupting kernel")
-                        self.km.interrupt_kernel()
-                        break
-                    else:
-                        raise TimeoutError("Cell execution timed out")
+                    self._handle_timeout()
+                    break
             else:
                 if msg['parent_header'].get('msg_id') == msg_id:
                     return msg
 
+    def _timeout_with_deadline(self, timeout, deadline):
+        if deadline is not None and deadline - monotonic() < timeout:
+            timeout = deadline - monotonic()
+
+        if timeout < 0:
+            timeout = 0
+
+        return timeout
+
+    def _passed_deadline(self, deadline):
+        if deadline is not None and deadline - monotonic() <= 0:
+            self._handle_timeout()
+            return True
+        return False
+
     def run_cell(self, cell, cell_index=0):
         parent_msg_id = self.kc.execute(cell.source)
         self.log.debug("Executing cell:\n%s", cell.source)
-        exec_reply = self._wait_for_reply(parent_msg_id, cell)
+        exec_timeout = self._get_timeout(cell)
+        deadline = None
+        if exec_timeout is not None:
+            deadline = monotonic() + exec_timeout
 
         cell.outputs = []
         self.clear_before_next_output = False
 
-        while True:
-            try:
-                # We've already waited for execute_reply, so all output
-                # should already be waiting. However, on slow networks, like
-                # in certain CI systems, waiting < 1 second might miss messages.
-                # So long as the kernel sends a status:idle message when it
-                # finishes, we won't actually have to wait this long, anyway.
-                msg = self.kc.iopub_channel.get_msg(timeout=self.iopub_timeout)
-            except Empty:
-                self.log.warning("Timeout waiting for IOPub output")
-                if self.raise_on_iopub_timeout:
-                    raise RuntimeError("Timeout waiting for IOPub output")
-                else:
-                    break
+        more_output = True
+        polling_exec_reply = True
+
+        while more_output or polling_exec_reply:
+            if polling_exec_reply:
+                if self._passed_deadline(deadline):
+                    polling_exec_reply = False
+                    continue
+
+                timeout = self._timeout_with_deadline(1, deadline)
+                exec_reply = self._poll_for_reply(parent_msg_id, cell, timeout)
+                if exec_reply is not None:
+                    polling_exec_reply = False
+
+            if more_output:
+                try:
+                    timeout = self.iopub_timeout
```

Comment:
So now with all the functionality in place, we likely need to address the fact that `iopub_timeout` no longer acts as it used to. In particular, since we aren't waiting for exec reply to fully complete we'll expect to timeout here often. Which means that `raise_on_iopub_timeout` will fire when users wouldn't expect.

@mpacer I think we should perhaps remove `raise_on_iopub_timeout` and `iopub_timeout` with this change and use `timeout = self._timeout_with_deadline(1, deadline)` for both conditional branches. Those other configurations just don't make any sense any more and are problematic if defined as they were in earlier versions. What's your thoughts? Even if we do a less severe change we'd need to address the change in timeout behavior somehow with this PR merging.

Reply:
Actually I tried to preserve that behavior, it won't hit the raise_on_iopub_timeout unless polling_exec_reply is false, and it's only false if we've passed the deadline or if exec_reply is not None (completed).

If we've passed the deadline we hit handle_timeout, which will either interrupt or throw. If it throws it won't reach this section, if it interrupts I suspect we would get an idle message?

Let me know if I got anything wrong here.

Reply:
Ahh yes you're correct, my end-of-conference brain power failed to read properly. Then maybe we should deprecate those fields in the next release (5.5.1) so they can be removed in 6.0? That would help cleanup this code and make it simpler to control in the future imo.

We should add a test for this case if it's not covered yet to ensure we won't timeout unexpectantly. That might be tricky to write with Mocks, so I can probably make that test if it's not clear how to set that up. After that I think I'm good to merge once @mpacer approves / has any further concerns resolved.

Reply:
Hey sorry, I was also gone for a week long work even this week.
Looks like I've got more merge conflicts to fix, I'll take a look Monday.

Reply:
Sorry there's been so many merge conflicts. I've been more careful to review / get more opinions on this PR before merging. Shouldn't be too many more conflicts. I asked @mpacer in person to take a look again when she's available so we don't keep this PR open for much longer.

Reply:
Yeah I understand, I think I managed to merge it in correctly now.

Reply:
I am not quite following the reasoning behind why the fields need to be deprecated. Could you lay out the situation where these fields cease to make sense?

Reply:
@MSeal probably has a better answer than me, my interpretation was that iopub_timeout exists to give a small window after receiving exec_reply (so we know all output should be produced) for the output to reach the iopub_channel (in case of network latency etc.). I think @MSeal might feel that since we are polling for this output more regularly (and not just after exec_reply) that the purpose of this flag has changed somewhat. Perhaps I shouldn't be using iopub_timeout at all unless exec_reply has finished, and use a constant timeout / deadline similar to _poll_for_reply. I could be misinterpreting @MSeal though so please correct me!

Even with all of that, I still think iopub_timeout is relevant post exec_reply being received, since you may still need to wait for output produced near the end of execution to be received.

Reply:
That's pretty much on point. It's odd the to have a post exec_reply response specific timeout in addition to an overall timeout when the two channels are being consumed in parallel.

I'll start a separate PR / Issue on the field independent of this PR to avoid leaving this open longer.

Reply:
@MSeal Thanks, what do you think about using a constant (like 1s) insead of the iopub_timeout in this PR for the poll to iopub_channel while polling_exec_reply=True, and then switching to iopub_timeout once exec_reply is received?

Reply:
Possibly that would be a way to approach it. In general I'm saying I think making it simpler and having 1 timeout would reduce confusion and be less nuanced.